`AdvancedTable`: Website documentation (#2628)
hashicorp/design-system: website/docs/components/table/advanced-table/partials/code/component-api.md

## Component API

The AdvancedTable component itself is where most of the options will be applied. However, the APIs for the child components are also documented here, in case a custom implementation is desired.

### AdvancedTable

<Doc::ComponentApi as |C|>
  <C.Property @name="<:body>" @type="named block">
    This is a named block where the content for the AdvancedTable body is rendered.
    <Doc::ComponentApi as |C|>
      <C.Property @name="[B].rowIndex" @type="number | string">
        The value of the index associated with the `@each` loop. Returns a number when there are no nested rows. Returns a string in the form `${parentIndex}.${childIndex}` when there are nested rows.
      </C.Property>
      <C.Property @name="[B].sortBy" @type="string">
        The value of the internal `sortBy` tracked variable.
      </C.Property>
      <C.Property @name="[B].sortOrder" @type="string">
        The value of the internal `sortOrder` tracked variable.
      </C.Property>
      <C.Property @name="[B].isExpanded" @type="boolean">

>         The value of the internal `isExpanded` tracked variable from the row if it has nested rows. Otherwise returns `undefined`.

Reviewer: ```suggestion
        Returns the value of the internal `isExpanded` tracked variable from the row if it has nested rows. Otherwise returns `undefined`.
```

  </C.Property>
  <C.Property @name="sortOrder" @type="string" @values={{array "asc" "desc" }} @default="asc">
    Use in conjunction with `sortBy`. If defined, indicates which direction the column should be pre-sorted in. If not defined, `asc` is applied by default.
  </C.Property>
  <C.Property @name="isSelectable" @type="boolean" @default="false">
    If set to `true`, creates a “multi-select” table which renders checkboxes in the table header and on the table rows enabling bulk interaction. Use in conjunction with `onSelectionChange` on the `Table` and `selectionKey` on each `Table::Tr`.<br><br>
    **Important**: AdvancedTable does **not** support having `isSelectable` true when there are nested rows.
  </C.Property>
  <C.Property @name="onSelectionChange" @type="function">
    Use in conjunction with `isSelectable` to pass a callback function to know the selection state. Must be used in conjunction with setting a `selectionKey` on each `Table::Tr`.
    <br /><br />
    When called, this function receives an object as argument, with different keys corresponding to different information:

>     <ul>

Author: Its used in the Table docs right now: https://helios.hashicorp.design/components/table?tab=code#table

It unfortunately can't be markdown because it is nested inside of an Ember component.

    <li>`selectedRowsKeys`: an array containing all the `@selectionKey`s of the selected rows in the table (an empty array is returned if no row is selected)</li>
    <li>`selectableRowsStates`: an array of objects corresponding to all the rows displayed in the table when the user changed a selection; each object contains the `@selectionKey` value for the associated row and its `isSelected` boolean state (if the checkbox is checked or not)<br><br>
    **Important**: the order of the rows in the array doesn’t necessarily follow the order of the rows in the table/DOM.</li>
    </ul>
  </C.Property>
  <C.Property @name="isStriped" @type="boolean" @default="false">
    Determines if even-numbered rows will have a different background color from odd-numbered rows.<br><br>
    **Important**: AdvancedTable does **not** support having `isStriped` true when there are nested rows.
  </C.Property>
  <C.Property @name="hasStickyHeader" @type="boolean" @default="false">
    Determines if the AdvancedTable has a sticky header.
  </C.Property>
  <C.Property @name="density" @type="enum" @values={{array "short" "medium" "tall" }} @default="medium">

>     If set, determines the density (height) of the body's rows.

Reviewer: ```suggestion
    If set, determines the density (height) of the body’s rows.
```

Reviewer: Just since you've used it elsewhere in here.

  </C.Property>
  <C.Property @name="isSelected" @type="boolean" @default="false">
    Sets the initial selection state for the row (used in conjunction with setting `isSelectable` on the `AdvancedTable`).
  </C.Property>
  <C.Property @name="selectionKey" @type="string | number">
    Required value to associate an unique identifier to each table row (used in conjunction with setting `isSelectable` on the `AdvancedTable` and returned in the `onSelectionChange` callback arguments). It’s required if `isSelectable={{true}}`.
  </C.Property>
  <C.Property @name="selectionAriaLabelSuffix" @type="string">
    Descriptive `aria-label` attribute applied to the checkbox used to select the row (used in conjunction with setting `isSelectable` on the `AdvancedTable`). The component automatically prepends “Select/Deselect” to the string, depending on the selection status. It’s required if `isSelectable={{true}}`.
  </C.Property>
  <C.Property @name="...attributes">
    This component supports use of [`...attributes`](https://guides.emberjs.com/release/in-depth-topics/patterns-for-components/#toc_attribute-ordering).
  </C.Property>
</Doc::ComponentApi>

### AdvancedTable::Th

> Note: This component is not eligible to receive interactions (e.g., it cannot have an `onClick` event handler attached directly to it). Instead, an interactive element should be placed _inside_ of the `AdvancedTable::Th` component.

Reviewer 2: [nit] Could also put this one in an info banner

```suggestion
!!! Info

Note: This component is not eligible to receive interactions (e.g., it cannot have an `onClick` event handler attached directly to it). Instead, an interactive element should be placed _inside_ of the `AdvancedTable::Th` component.
!!!
```

If the `Th` component is passed as the first cell of a body row, `role="rowheader"` is automatically applied for accessibility purposes.

<Doc::ComponentApi as |C|>
  <C.Property @name="align" @type="enum" @values={{array "left" "center" "right" }} @default="left">
    Determines the horizontal content alignment (sometimes referred to as text alignment) for the column header.
  </C.Property>
  <C.Property @name="scope" @type="string" @values={{array "col" "row" }} @default="col">
    If used as the first item in a table body’s row, `scope` should be set to `row` for accessibility purposes. Note: you only need to manually set this if you’re creating a custom table using the child components; if you use the standard invocation for the table, this scope is already provided for you.
  </C.Property>
  <C.Property @name="width" @type="string" @valueNote="Any valid CSS">
    If set, determines the column’s width.
  </C.Property>
  <C.Property @name="tooltip" @type="string">

>     Text string which will appear in the tooltip (see [`Tooltip`](/components/tooltip) for details). May contain basic HTML tags for formatting text such as `strong` and `em` tags. Not intended for multi-paragraph text or other more complex content. May not contain interactive content such as links or buttons. The `placement` and `offset` are automatically set and can’t be overwritten.

Reviewer 2: Same suggestion as above

```suggestion
    Text string which will appear in the [Tooltip](/components/tooltip). May contain basic HTML tags for formatting text such as `strong` and `em` tags. Not intended for multi-paragraph text or other more complex content. May not contain interactive content such as links or buttons. The `placement` and `offset` are automatically set and can’t be overwritten.
```

        Text string which will appear in [`Tooltip`](/components/tooltip). May contain basic HTML tags for formatting text such as `strong` and `em` tags. Not intended for multi-paragraph text or other more complex content. May not contain interactive content such as links or buttons. The `placement` and `offset` are automatically set and can’t be overwritten.
      </C.Property>
    </Doc::ComponentApi>
  </C.Property>
  <C.Property @name="sortBy" @type="string">
    If defined, the value should be set to the key of the column that should be pre-sorted.
  </C.Property>
  <C.Property @name="sortOrder" @type="string" @values={{array "asc" "desc" }} @default="asc">
    Use in conjunction with `sortBy`. If defined, indicates which direction the column should be pre-sorted in. If not defined, `asc` is applied by default.
  </C.Property>
  <C.Property @name="isSelectable" @type="boolean" @default="false">
    If set to `true`, creates a “multi-select” table which renders checkboxes in the table header and on the table rows enabling bulk interaction. Use in conjunction with `onSelectionChange` on the `Table` and `selectionKey` on each `Table::Tr`.<br><br>

>     **Important**: AdvancedTable does **not** support having `isSelectable` true when there are nested rows.

Reviewer 3: ```suggestion
    **Important**: Advanced Table does **not** support having `isSelectable` true when there are nested rows.
```
[nit] aligning naming to other instances
